### PR TITLE
Db structure db - semantic ai

### PR DESCRIPTION
**Pull request by JeyachandranV**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -112,6 +125,45 @@ connection = Sharepoint(client_id=CLIENT_ID,
...
+import semantic_ai
+```
+
+### 2. The database connection  
+
+#### Sqlite:
+file_path= <database_file_path>
+
+sql = Sqlite(sql_path=file_path)
+
```
Comment (Balamurugan35) [inline README.md:144]: import the Sqlite class

File: README.md [hunk 2]
```diff
@@ -112,6 +125,45 @@ connection = Sharepoint(client_id=CLIENT_ID,
...
+
+#### Mysql:
+```
+sql = Mysql(
```
Comment (Balamurugan35) [inline README.md:148]: import Mysql class

File: docs/source/connectors/mysql.rst [hunk 3]
```diff
@@ -0,0 +1,64 @@
+MYSQL
+======
+
+.. figure:: https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRmvZum7UQzH7A7I6M-f4Rjdth_UEf4i6_0SA&usqp=CAU
```
Comment (Balamurugan35) [inline docs/source/connectors/mysql.rst:4]: Don't use other public image url. Use from our repo url. Refer previous docs

File: docs/source/connectors/mysql.rst [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    MYSQL_PORT='<port>'
+
+
+.. code-block:: python
+
+    from semantic_ai.connectors import Mysql
+
+        conn = Mysql(
+                    host='',
+                    user='',
```
Comment (Balamurugan35) [inline docs/source/connectors/mysql.rst:40]: don't put empty qutoes. Give name of the variable like this ('host', 'port')

File: docs/source/connectors/sqlite.rst [hunk 5]
```diff
@@ -0,0 +1,54 @@
+SQLITE
+======
+
+.. figure:: https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQDlMIPeUYeHYEWmS0hX5Qt3WrbA-wVvk2nrw&usqp=CAU
```
Comment (Balamurugan35) [inline docs/source/connectors/sqlite.rst:4]: Don't use other public image url. Use from our repo url. Refer previous docs

File: docs/source/connectors/sqlite.rst [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+.. code-block:: python
+
+    from semantic_ai.connectors import Sqlite
+
+    file_path = f'<DB_PATH>'
+    sqlite = Sqlite(sql_path=file_path)
+    conn = await sqlite.connect_db()
+
+**To create natural language processing for normal text**
```
Comment (Balamurugan35) [inline docs/source/connectors/sqlite.rst:38]: Change the topic  "To generate the sql query from NLP text"

File: docs/source/nlp/promp.rst [hunk 7]
```diff
@@ -0,0 +1,34 @@
+Natural Language Processing
+======
+
+.. figure:: https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcSuqJW1cosMDRjkXvBnJ8eWZp0q0xT2nG5tpA&usqp=CAU
```
Comment (Balamurugan35) [inline docs/source/nlp/promp.rst:4]: Don't use other public image url. Use from our repo url. Refer previous docs

File: semantic_ai/__init__.py [hunk 8]
```diff
@@ -132,6 +134,36 @@ async def index(settings: Settings | None = None):
         raise ValueError(f"Give valid credentials or load properly environment variable.")
 
 
+async def db_search(query, settings: Settings | None = None):
+    settings = await _load_settings(settings)
+    if not settings.connector_type:
+        raise "The connector type does not supporting"
```
Comment (Balamurugan35) [inline semantic_ai/__init__.py:140]: Change the error text 'connector type should not be empty'

File: semantic_ai/connectors/microsoft/mysql.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+from semantic_ai.connectors.base import BaseSqlConnector
+from langchain.utilities import SQLDatabase
+from fastapi import HTTPException, status
+
+from semantic_ai.utils import sync_to_async
+
+logger = logging.getLogger(__name__)
+
+
+SQL_RESPONSE_TEMPLATE = """Generate natural language information based {question} and {response}.
```
Comment (Balamurugan35) [inline semantic_ai/connectors/microsoft/mysql.py:16]: This is common prompt for all sturcture

File: semantic_ai/connectors/microsoft/mysql.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        except Exception as ex:
+            logger.error('Search query exception => ', exc_info=ex)
+            raise HTTPException(
+                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
+                detail="Sorry! Internal server error while executing query."
+            )
+
+    async def mysql_client(self):
+        try:
+            return mysql.connector.connect(user=self.user,
```
Comment (Balamurugan35) [inline semantic_ai/connectors/microsoft/mysql.py:45]: convert into async

File: semantic_ai/connectors/microsoft/mysql.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+                logger.error(err)
+                raise err
+
+    async def execution(self, data: dict):
+        try:
+            query = data.get('SQLQuery')
+            question = data.get('Question')
+            currency = data.get('Currency')
+            conn = await self.mysql_client()
+            curr = conn.cursor()
```
Comment (Balamurugan35) [inline semantic_ai/connectors/microsoft/mysql.py:67]: convert into async

File: semantic_ai/connectors/microsoft/mysql.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            curr.execute(query)
```
Comment (Balamurugan35) [inline semantic_ai/connectors/microsoft/mysql.py:68]: convert into async

File: semantic_ai/connectors/microsoft/mysql.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+            labels = [description[0] for description in curr.description]
+            response = curr.fetchall()
```
Comment (Balamurugan35) [inline semantic_ai/connectors/microsoft/mysql.py:70]: convert into async

File: semantic_ai/examples/test_mysql_search.py [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+from semantic_ai.connectors import Mysql
+
+from semantic_ai.nlp.prompt import Prompt
+
+sql = Mysql(
+    host='',
+    user='',
+    password='',
+    database='',
+    port=""  # 3306 is default port
```
Comment (Balamurugan35) [inline semantic_ai/examples/test_mysql_search.py:13]: don't put empty string

File: semantic_ai/examples/test_sqlite_search.py [hunk 15]
```diff
@@ -0,0 +1,17 @@
+import asyncio
+import json
+
+from semantic_ai.connectors import Sqlite
+
+from semantic_ai.nlp.prompt import Prompt
+
+file_path = f''
```
Comment (Balamurugan35) [inline semantic_ai/examples/test_sqlite_search.py:8]: don't put empty string

File: semantic_ai/examples/test_mysql_search.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+)
+cur = asyncio.run(sql.connect_db())
+nlp_to_sql = asyncio.run(Prompt().nlp_to_sql(data_base=cur, normal_text=""))
+data = json.loads(nlp_to_sql)
+result = asyncio.run(sql.execution(data))
```
Comment (Balamurugan35) [inline semantic_ai/examples/test_mysql_search.py:18]: use await instead of asyncio

File: semantic_ai/nlp/prompt.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+    async def get_llm_chain(self, prompt: str = None):
+        openai_llm = Openai()
+        if not prompt:
+            prompt = self.prompt
+        return LLMChain(llm=await openai_llm.llm_model(), prompt=prompt)
+
+    async def get_db_context(self, query: str, db):
+        openai_llm = Openai()
+        db_chain = SQLDatabaseChain.from_llm(openai_llm, db, verbose=True)
+        db_context = db_chain(query)
```
Comment (Balamurugan35) [inline semantic_ai/nlp/prompt.py:53]: convert into async

**Review by Balamurugan35 — COMMENTED** [review]
Refer previous release

File: semantic_ai/nlp/prompt.py [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    async def get_db_context(self, query: str, db):
+        db_chain = SQLDatabaseChain.from_llm(await self.openai_llm.llm_model(), db, verbose=True)
+        db_context = await sync_to_async(db_chain, query)
+        db_context = db_context['result'].strip()
+        return db_context
+
+    async def nlp_to_sql(self, data_base: SQLDatabase, normal_text: str, prompt: str = None):
+        try:
+            chain = await self.get_llm_chain(prompt)
+            print(normal_text)
```
Comment (SyedHabibM) [inline semantic_ai/nlp/prompt.py:61]: Remove this

File: semantic_ai/nlp/prompt.py [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class Prompt:
+
+    def __init__(self):
+        self.prompt = PromptTemplate(
+            input_variables=["input", "table_info", "dialect"],
+            template=SQL_DEFAULT_TEMPLATE
+        )
+        self.openai_llm = Openai(model_name_or_path="gpt-3.5-turbo-1106")
```
Comment (SyedHabibM) [inline semantic_ai/nlp/prompt.py:45]: Why this hard coded here?!

File: semantic_ai/llm/llama.py [hunk 20]
```diff
@@ -32,50 +30,53 @@ def __init__(self,
...
+                pad_token_id=tokenizer.eos_token_id,
+                inject_fused_attention=False,
+                early_stopping=True,
+                use_triton=False
+            )
+            gptq_model = torch.compile(gptq_model)
+            # gptq_model = exllama_set_max_input_length(gptq_model, 4096)
+            gptq_model = exllama_set_max_input_length(gptq_model, 8192)
+            gptq_model.eval()
+            print(f"Model loaded on {device}")
```
Comment (SyedHabibM) [inline semantic_ai/llm/llama.py:57]: Remove this

File: semantic_ai/connectors/sqlite.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+                }
+            ]
+            openai_res = await self.client.chat.completions.create(
+                model="gpt-3.5-turbo-1106",
+                messages=messages
+            )
+            llm_result = openai_res
+            if not llm_result:
+                return resp
+            resp['result'] = f'{llm_result.choices[0].message.content}'
```
Comment (SyedHabibM) [inline semantic_ai/connectors/sqlite.py:66]: Avoid index out of range error

File: semantic_ai/connectors/mysql.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+
+            template = SQL_RESPONSE_TEMPLATE.format(question=question, response=response)
+            messages = [
+                {
+                    "role": "system",
+                    "content": template
+                }
+            ]
+            openai_res = await self.client.chat.completions.create(
+                model="gpt-3.5-turbo-1106",
```
Comment (SyedHabibM) [inline semantic_ai/connectors/mysql.py:89]: Why this hard coded here?!

**Review by SyedHabibM — CHANGES_REQUESTED** [review]

File: semantic_ai/nlp/prompt.py (hunk 18 above)
Comment (Balamurugan35) [reply semantic_ai/nlp/prompt.py]: Removed

File: semantic_ai/nlp/prompt.py (hunk 19 above)
Comment (Balamurugan35) [reply semantic_ai/nlp/prompt.py]: This is the default model. Am changed into the as a params  for model. We can change model thourgh params.

File: semantic_ai/llm/llama.py (hunk 20 above)
Comment (Balamurugan35) [reply semantic_ai/llm/llama.py]: Removed

File: semantic_ai/connectors/sqlite.py (hunk 21 above)
Comment (Balamurugan35) [reply semantic_ai/connectors/sqlite.py]: reformat this line

File: semantic_ai/connectors/mysql.py (hunk 22 above)
Comment (Balamurugan35) [reply semantic_ai/connectors/mysql.py]: This is the default model. Am changed into the as a params for model. We can change model thourgh params.

**Review by SyedHabibM — APPROVED** [review]

**Pull request merged**